### ISSUE TITLE
Adding library info to symbol name

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#278]
In some programs that have a lot of libraries are linked, it's difficult to identify which libraries contain each function.  So I think it'd be useful to show such library information to each symbol name.

What about adding a new option may be called `--libinfo` and makes it add symbol name suffix `@<libname>`. For example, `malloc()` can be shown as `malloc@libc.so.6()`.  It can be especially useful when tracing pre-built binary using `--nest-libcall`.

What do you think?

**Comment by namhyung** [comment]
But PLT symbols belong to callers, not a library it defines.

**Comment by honggyukim** [comment]
Hmm.. You're right. What about checking the target address and then compare it with maps info? It cannot be possible at the first call, but possible at the second call with resolved GOT table. Of course, tracing from the first call may be the problem.

**Comment by namhyung** [comment]
Symbol resolution is done during replay so it has nothing to deal with GOT.  The problem is the order of resolution (library search).  But I think it's ok for most cases since (public) functions are usually defined in a single library.

**Comment by namhyung** [comment]
I've uploaded the PoC code to [review/libname-support-v1](https://github.com/namhyung/uftrace/compare/review/libname-support-v1).

It'll show name of library which defines the current function (for PLT functions):

```
$ uftrace --libname foobar
# DURATION     TID     FUNCTION
   1.365 us [  7129] | __monstartup();
   1.208 us [  7129] | __cxa_atexit@libc-2.26.so();
            [  7129] | main@foobar() {
            [  7129] |   foo@foobar() {
   0.143 us [  7129] |     bar@foobar();
   0.735 us [  7129] |   } /* foo@foobar */
  74.593 us [  7129] |   pthread_create@libpthread-2.26.so();
            [  7129] |   pthread_join@libpthread-2.26.so() {
            [  7133] | foo@foobar() {
   0.192 us [  7133] |   bar@foobar();
   1.471 us [  7133] | } /* foo@foobar */
 239.315 us [  7129] |   } /* pthread_join@libpthread-2.26.so */
 316.967 us [  7129] | } /* main@foobar */
```

**Comment by honggyukim** [comment]
Thanks but I found a segfault when running it with a qml program `uftrace --libname qml hello.qml`.  Please take a look at https://github.com/namhyung/uftrace/commit/99d88fc814d29d10f9137444d88f1b253f952eb1#r26586121.

And I'm just wondering if the below output is correct.
```
$ uftrace --libname tests/t-fork
# DURATION     TID     FUNCTION
   0.329 us [ 19802] | __monstartup();
   0.160 us [ 19802] | __cxa_atexit@libstdc++.so.6.0.22();
            [ 19802] | main@t-fork() {
  55.128 us [ 19802] |   fork@libpthread-2.24.so();
            [ 19802] |   wait@libpthread-2.24.so() {
            [ 19806] |   } /* fork@libpthread-2.24.so */
            [ 19806] |   a@t-fork() {
            [ 19806] |     b@t-fork() {
            [ 19806] |       c@t-fork() {
   0.100 us [ 19806] |         getpid@librt-2.24.so();
   2.209 us [ 19806] |       } /* c@t-fork */
   2.381 us [ 19806] |     } /* b@t-fork */
   2.540 us [ 19806] |   } /* a@t-fork */
   3.606 us [ 19806] | } /* main@t-fork */
 208.730 us [ 19802] |   } /* wait@libpthread-2.24.so */
            [ 19802] |   a@t-fork() {
            [ 19802] |     b@t-fork() {
            [ 19802] |       c@t-fork() {
   0.094 us [ 19802] |         getpid@librt-2.24.so();
   0.500 us [ 19802] |       } /* c@t-fork */
   0.672 us [ 19802] |     } /* b@t-fork */
   0.826 us [ 19802] |   } /* a@t-fork */
 266.595 us [ 19802] | } /* main@t-fork */
```
I don't know why `fork()` is from `libpthread-2.24.so` and `getpid()` is from `librt-2.24.so`.

**Comment by namhyung** [comment]
The libpthread has a stub function for `fork()` which calls `__libc_fork()` internally:

```
$ nm --dynamic /usr/lib/libpthread-2.26.so | grep fork
0000000000010840 T fork
0000000000010840 T __fork
                 U __libc_fork
                 U __libc_vfork
000000000000ee80 T pthread_atfork
                 U __register_atfork
0000000000010850 i vfork
0000000000010860 i __vfork


$ objdump -d --start-address=0x10840 --stop-address=0x10850 /usr/lib/libpthread-2.26.so

/usr/lib/libpthread-2.26.so:     file format elf64-x86-64


Disassembly of section .text:

0000000000010840 <__fork@GLIBC_2.2.5>:
   10840:	ff 25 ea 85 20 00    	jmpq   *0x2085ea(%rip)        # 218e30 <__libc_fork@GLIBC_PRIVATE>
   10846:	66 2e 0f 1f 84 00 00 	nopw   %cs:0x0(%rax,%rax,1)
   1084d:	00 00 00 
```

**Comment by namhyung** [comment]
As searching library is done by its address, the result might be varying if more than one library define the same function.

**Comment by namhyung** [comment]
Pushed [review/libname-support-v2](https://github.com/namhyung/uftrace/compare/review/libname-support-v2).

Changelog:
* fix crash when main map was returned
* show libname only for PLT functions
* add IFUNC symbols

```
$ uftrace --libname abc-pg
# DURATION     TID     FUNCTION
   1.110 us [ 20070] | __monstartup();
   0.700 us [ 20070] | __cxa_atexit@libc-2.26.so();
            [ 20070] | main() {
            [ 20070] |   a() {
            [ 20070] |     b() {
            [ 20070] |       c() {
   0.655 us [ 20070] |         getpid@libc-2.26.so();
   1.500 us [ 20070] |       } /* c */
   1.875 us [ 20070] |     } /* b */
   2.118 us [ 20070] |   } /* a */
   2.460 us [ 20070] | } /* main */
```

**Comment by honggyukim** [comment]
Thanks but I got another segfault. Please see the details below:
```
$ uftrace record qml hello.qml

$ cgdb --args uftrace replay --libname --no-pager
                    ...
Program received signal SIGSEGV, Segmentation fault.
0x000055555556d304 in print_graph_rstack (opts=0x7fffffffdb80, task=<optimized out>, handle=0x7fffffffd0c0) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:613
                    ...
583│ static int print_graph_rstack(struct ftrace_file_handle *handle,
584│                               struct ftrace_task_handle *task,
585│                               struct opts *opts)
586│ {
                    ...
613├───────> if (opts->libname && sym->type == ST_PLT) {
614│                 struct uftrace_session *s;
615│
616│                 s = find_task_session(sessions, task->tid, rstack->time);
617│                 if (s) {
618│                         map = find_symbol_map(&s->symtabs, symname);
619│                         if (map && map != MAP_MAIN)
620│                                 libname = basename(map->libname);
621│                 }
622│         }

(gdb) p sym
$1 = (struct sym *) 0x0

(gdb) bt
#0  0x000055555556d304 in print_graph_rstack (opts=0x7fffffffdb80, task=<optimized out>, handle=0x7fffffffd0c0) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:613
#1  command_replay (argc=<optimized out>, argv=<optimized out>, opts=<optimized out>) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:985
#2  0x000055555555bdd5 in main (argc=4, argv=0x7fffffffde68) at /home/honggyu/src/uftrace/git/clean/uftrace/uftrace.c:1002
```

**Comment by honggyukim** [comment]
One more thing is that I think it would be better to
- show libname for every function
- distinguish PLT functions to be `getpid@plt,libc-2.26.so()` rather than `getpid@libc-2.26.so`

Please let me know how you think about it.

**Comment by namhyung** [comment]
Fixed in [review/libname-support-v3](https://github.com/namhyung/uftrace/compare/review/libname-support-v3).  I also added `module` field to show library/exename for every function.

**Comment by honggyukim** [comment]
Hmm.. It crashes.
```
$ uftrace record --nest-libcall ./qml hello.qml

$ cgdb --args uftrace replay -f +module --no-pager
        ...
Program received signal SIGSEGV, Segmentation fault.
print_module (fd=<optimized out>) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:111
        ...
  95│ static void print_module(struct field_data *fd)
  96│ {
  97│         struct ftrace_task_handle *task = fd->task;
  98│         struct fstack *fstack = fd->fstack;
  99│         uint64_t timestamp = task->timestamp;
 100│         struct uftrace_session *s;
 101│         struct uftrace_mmap *map;
 102│         char *modname = "[unknown]";
 103│
 104│         s = find_session(&task->h->sessions, task->tid, timestamp);
 105│         if (s == NULL)
 106│                 s = find_session(&task->h->sessions, task->t->pid, timestamp);
 107│         if (s == NULL)  /* for fork/vfork() */
 108│                 s = find_session(&task->h->sessions, task->t->ppid, timestamp);
 109│
 110│         if (s) {
 111├───────────────> map = find_map(&s->symtabs, fstack->addr);
 112│                 if (map == MAP_MAIN)
 113│                         modname = basename(s->exename);
 114│                 else if (map == MAP_KERNEL)
 115│                         modname = "[kernel]";
 116│                 else if (map)
 117│                         modname = basename(map->libname);
 118│         }
 119│
 120│         pr_out("%*.*s", 16, 16, modname);
 121│ }

(gdb) p fstack
$2 = (struct fstack *) 0x0

(gdb) bt
#0  print_module (fd=<optimized out>) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:111
#1  0x0000555555578c62 in print_field_data (output_fields=output_fields@entry=0x5555557ac6c0 <output_fields>, fd=fd@entry=0x7fffffffcfb0, space=space@entry=1) at /home/honggyu/src/uftrac
e/git/clean/uftrace/utils/field.c:34
#2  0x000055555556c526 in print_field (task=task@entry=0x555555d550e0, fstack=fstack@entry=0x0, arg=arg@entry=0x1) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:221
#3  0x000055555556d3bd in print_graph_rstack (opts=0x7fffffffdb80, task=<optimized out>, handle=0x7fffffffd0c0) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:855
#4  command_replay (argc=<optimized out>, argv=<optimized out>, opts=<optimized out>) at /home/honggyu/src/uftrace/git/clean/uftrace/cmd-replay.c:1023
#5  0x000055555555be65 in main (argc=5, argv=0x7fffffffde68) at /home/honggyu/src/uftrace/git/clean/uftrace/uftrace.c:1002
```

**Comment by honggyukim** [comment]
In the recorded data, some functions do not have symbol name in the original trace. It seems that missing symbols can make such segfault.
```
$ uftrace replay -t 20ms
        ...
            [ 19287] |   QCoreApplication::exec() {
            [ 19287] |     QCoreApplication::notifyInternal2() {
            [ 19287] |       QCoreApplication::notify() {
            [ 19287] |         <7f06e8a6a875>() {
            [ 19287] |           <7f06e8a58e9a>() {
            [ 19287] |             <7f06e89d2d1b>() {
            [ 19287] |               <7f06e89d1b5a>() {
            [ 19287] |                 <7f06e8a5cd8f>() {
            [ 19287] |                   <7f06e89e51d5>() {
            [ 19287] |                     <7f06e899b855>() {
  20.517 ms [ 19287] |                       <7f06e899b2be>();
  20.518 ms [ 19287] |                     } /* <7f06e899b855> */
  20.518 ms [ 19287] |                   } /* <7f06e89e51d5> */
  20.530 ms [ 19287] |                 } /* <7f06e8a5cd8f> */
  28.214 ms [ 19287] |               } /* <7f06e89d1b5a> */
  28.216 ms [ 19287] |             } /* <7f06e89d2d1b> */
  28.216 ms [ 19287] |           } /* <7f06e8a58e9a> */
  28.217 ms [ 19287] |         } /* <7f06e8a6a875> */
  28.218 ms [ 19287] |       } /* QCoreApplication::notify */
  28.218 ms [ 19287] |     } /* QCoreApplication::notifyInternal2 */
        ...
```

**Comment by namhyung** [comment]
I guess it's an event record.  Does the [following patch](https://github.com/namhyung/uftrace/files/1599423/replay-skip-event-for-module-field.patch.txt) help?

```diff
diff --git a/cmd-replay.c b/cmd-replay.c
index 836aa06..4576122 100644
--- a/cmd-replay.c
+++ b/cmd-replay.c
@@ -101,6 +101,12 @@ static void print_module(struct field_data *fd)
        struct uftrace_mmap *map;
        char *modname = "[unknown]";
 
+       /* for EVENT or LOST record */
+       if (fstack == NULL) {
+               pr_out("%*s", 16, "");
+               return;
+       }
+
        s = find_session(&task->h->sessions, task->tid, timestamp);
        if (s == NULL)
                s = find_session(&task->h->sessions, task->t->pid, timestamp);
```

**Comment by honggyukim** [comment]
You're right. Now it's fine with your patch above. Thanks!

**Comment by namhyung** [comment]
Merged 31db867
